dask/array/random.py

```diff
@@ -69,10 +109,15 @@ def _wrap(self, func, *args, **kwargs):
         name = 'da.random.{0}-{1}'.format(func.__name__, token)
         keys = product([name], *([range(len(bd)) for bd in chunks]
                                + [[0]] * len(extra_chunks)))
-        vals = ((_apply_random, func.__name__, seed, size, args, kwargs)
-                for seed, size in zip(seeds, sizes))
-        dsk = dict(zip(keys, vals))
 
+        vals = []
+        for seed, size, slc in zip(seeds, sizes, slices):
+            arg = [ar[slc] if isinstance(ar, (np.ndarray, Array)) else ar for ar
+                   in new_args]
```

Comment:
Ok, both issues should be solved now.

I tried to figure out how to do that with `atop` as you suggested, however in the end even the skeleton of that solution seemed overly complex, so I just did it directly.

Seems to work as far as I can tell.

The only gripe I have with it is that identical (in values) numpy arrays do not get identical tokens: